### ISSUE TITLE
CMS: Do we support negative counts and/or is computing inner products broken?

### ISSUE THREAD

**Issue by miguno** [twitter/algebird#393]
## Background

I came across this question while debugging a failing test (computing the inner product) for CMS with the proposed new hashing code for `BigInt` discussed at #392.  The failing test in question is ["estimate inner products"](https://github.com/twitter/algebird/blob/develop/algebird-test/src/test/scala/com/twitter/algebird/CountMinSketchTest.scala#L198).

I noticed that with the old/legacy CMS hashing for BigInt's -- and because they hash in the same way this means the issue I am talking about is also relevant for `CMS[Long]`, `CMS[Int]`, etc. -- that computing inner products (as part of the test above) returned `Approximate` instances with a negative minimum value:

```scala
case class Approximate[N](min: N, estimate: N, max: N, probWithinBounds: Double)(implicit val numeric: Numeric[N]) { ...}
```

Example using current code and current BigInt hashing:

```
Approximate(-2544,24850,24850,0.9999999999)  // <<< note the negative minimum of `-2544`
```

This also meant that the test ended up computing weird (?) error bounds because "minus a negative value" is effectively addition.  But the test passed, and has done so for I think many months.

```
approxA.min:  -2544
approx:       24850  (approx = approxA.estimate)
exact:         1192
estimError:   23658  (estimationError = approx - exact)
maxError:     27394  (maxError = approx - approxA.min)  // <<< IMHO maxError should rather be smaller than approx, not larger!
```

> Btw, the same negative value for `Approximate.min` happens for `CMS[Long]`, for instance.

When testing the proposed Murmur3-based hashing for `CMS[BigInt]`, the same tests fails, because here Approximate.min happened to be a positive value.

> Note: I am not claiming (yet?) that the values are negative all the time etc. -- I can only confirm that they happened to be negative in my limited testing.  Same thing for the "positive" values for the minimums returned with the new hashing code.

```
Approximate(6270,33664,33664,0.9999999999) // <<< min is positive, `6270`

approxA.min: 6270
approx:       33664  (approxA.estimate)
exact:         1305
estimError:   32359  (approx - exact)
maxError:     27394  (approx - approxA.min) // <<< this makes more sense, but now the test fails!
```

At this point I got confused.

I checked the [inner product definition](https://github.com/twitter/algebird/blob/develop/algebird-core/src/main/scala/com/twitter/algebird/CountMinSketch.scala#L484) of CMSInstance[K]:

```scala
  def innerProduct(other: CMS[K]): Approximate[Long] = {
    other match {
      case other: CMSInstance[_] =>
        require((other.depth, other.width) == (depth, width), "Tables must have the same dimensions.")

        def innerProductAtDepth(d: Int) = (0 to (width - 1)).iterator.map { w =>
          countsTable.getCount(d, w) * other.countsTable.getCount(d, w)
        }.sum

        val est = (0 to (depth - 1)).iterator.map { innerProductAtDepth }.min
        Approximate(est - (eps * totalCount * other.totalCount).toLong, est, est, 1 - delta)
      case _ => other.innerProduct(this)
    }
  }
```

The relevant line is:

```scala
        Approximate(est - (eps * totalCount * other.totalCount).toLong, est, est, 1 - delta)
```

Here, `est - (eps * totalCount * other.totalCount).toLong` computes the minimum, and this computation can result in negative values.

**Hence my question: Do we actually support negative values in CMS?**

I don't think we do, even though the original CMS paper does allow for CMS that support positive counts only as well as CMS that support positive and negative counts.  (Off the hip one reason why I think we do not support negative counts is that this might break the algebraic properties we need for CMS in Algebird?) If we actually *do* support negative counts, then for instance @avibryant's proposed code in #390 needs further adjustment as his List-based optimization only supports adding items but not removing any, nor does it support negative counts in general.

## A fix?

At this point I wondered whether my new hashing code (using Murmur3) is wrong, or whether we have a legitimate bug in the way we compute inner products in CMS, or even both issues together.

Out of curiosity I went ahead and tested a modification of `innerProduct()` where I set the minimum value to 0 if the normal computation produces a negative value:

```scala
        val minimum = Math.min(est - (eps * totalCount * other.totalCount).toLong, 0)
        Approximate(minimum, est, est, 1 - delta)
```

And surprise:  Now _all_ tests in Algebird passed again, and they have been passing consistently (I re-ran the full test suite a couple of times).  And when I say "all" tests passed then this included a) my proposed hashing changes for CMS[BigInt] in #392 -- where the failing test described above caused me to doubt my code changes -- as well as b) the fickle CMS params round-trip test that @johnynek flagged as buggy in #388.  Uh?

## TL;DR

I'd appreciate any feedback in this matter.  Are we supposed to support negative counts?  Is the current implementation of inner products correct or buggy?  And even though the next question is harder to answer at the moment: How could a change of the CMS hashing algo for `BigInt` trigger a bug in the inner product computation but only there, and not e.g. in "normal" frequency estimation aka point queries?

Another reason for me to ask this question is that @avibryant is apparently very much interested in leveraging and tuning the CMS code because he works on clustering based on inner products of CMS, so any bug we might have here should be ironed out.

PS:  I also begin to get the feeling the CMS code needs more testing to ensure it works as expected, given the recent issues with it...

**Comment by jnievelt** [comment]
I believe it is not supported.  See:

https://github.com/twitter/algebird/blob/cdaa2b2fe735093d1cd45dc63783f4253d3e5cca/algebird-core/src/main/scala/com/twitter/algebird/CountMinSketch.scala#L500

Granted, it doesn't seem to be enforced as exhaustively as it could be.  For example, I believe it would be possible to build a `CMSInstance.CountsTable` with negative counts and build a `CMSInstance` from that.

However, it seems that before asking whether this is supported, we should look in more depth at what the tests are doing.  First of all, are we certain that the input to these tests have positive counts?  Furthermore, can we convince ourselves from a mathematical perspective that the formula for lower bound should never be negative given positive input counts?

If we instead convince ourselves that the formula can be negative, and we maintain non-support for negative counts, using a minimum lower bound of 0 sounds reasonable to me.  Alternatively, we would need to relax our requirement that the lower bound is non-negative.

**Comment by miguno** [comment]
All good points, Joe, and thanks for the quick feedback!

> First of all, are we certain that the input to these tests have positive counts?

Yes, the inputs to the inner product test are always positive:

```scala
      val totalCount = 5234
      val range = 1390
      val data1 = (0 to (totalCount - 1)).map { _ => RAND.nextInt(range) }.toK[K]
      val data2 = (0 to (totalCount - 1)).map { _ => RAND.nextInt(range) }.toK[K]
```

Example (same for `data2`):

```scala
scala> val data1 = (0 to (totalCount - 1)).map { _ => RAND.nextInt(range) }
scala> data1.filter(_ < 0).size
res2: Int = 0  // <<< no negative input values
```

> Furthermore, can we convince ourselves from a mathematical perspective that the formula for lower bound should never be negative given positive input counts?

From what I understand, yes:  If the input counts are positive, then there should never be negative results.

1. The counting table is initialized with all zeroes.
2. Assuming positive input counts, the `update` function in CMS (in Algebird: `+` and `++`) will always result in monotonically increasing "output" counts (`>=` or `>` doesn't matter here).  Basically, CMS does "slotCount += newCountValue".  (See page 6, "Update procedure" in the original [CMS paper](http://dimacs.rutgers.edu/~graham/pubs/papers/cm-full.pdf).)
3. The formulae for point queries and inner products (in Algebird: `frequency` and `innerProduct`), respectively, work similarly.  They essentially compute the L1-norm of the "vectors" that make the counting table, and pick the minimum -- which given (1) and (2) should always be `>= 0`.  In the case of inner products we do `(a_i * b_i).sum`, which again should always be `>= 0` (see page 8 in the paper, section 4.2 "Inner Product Query").

I also noticed how a helper function of `CMSInstance[K].++` uses the same approach that I used for preventing negative values:

```scala
  private def makeApprox(est: Long): Approximate[Long] = {
    if (est == 0L) Approximate.exact(0L)
    else {
      val lower = math.max(0L, est - (eps * totalCount).toLong) // <<< see here
      Approximate(lower, est, est, 1 - delta)
    }
  }
```

Which leads me to my follow up question:  Where can I find the definition of the formula we use to compute `Approximate.min`?  That is:

* `Approximate.min = math.max(0L, est - (eps * totalCount).toLong)` for `++`.
* `Approximate.min = math.max(0L, est - (eps * totalCount * other.totalCount).toLong)` for `innerProduct`.
    * The current code still uses the "negative values possible" variant: `Approximate.min = est - (eps * totalCount * other.totalCount).toLong` for `innerProduct`.

I can't seem to find it in the original paper, and at the moment they are my only candidates which may cause these negative values for `Approximate.min`.  In other words, unless I am mistaken I'd argue that the only place in the code -- based on my quick search -- where such negative results may happen is in the `innerProduct` function.  The `++` function already enforces a lower bound of `0`.

The reason I am asking this question is that, similar to your comment, maybe our tests are not working as intended -- e.g. maybe we have a faulty way to compute the `maxError` in the test or sth like that.

--

Also, `*blush*`, I noticed I made a mistake in my "fix" above -- I used `math.min(0, ...)` where I should have used `math.max(0, ...)`.  The incorrect use of `math.min()` fixed the issue IMHO only coincidentally.

Applying the mathematically more appropriate fix -- i.e. `math.max()` -- does not fix the issue.  Yes, we do not see negative values for `Approximate.min` anymore, and all the tests for the legacy code continue to pass successfully.  However, the inner product test is now failing again for the new code that uses Murmur3-based hashing for `BigInt]`.

**Comment by jnievelt** [comment]
Ah, of course.  `frequency` is a special case of `innerProduct` (i.e., `innerProduct` of a `CMS` with a `CMSItem`).  And it's pretty easy to see how `frequency` could give a negative lower bound.  So my guess is that we'd want to `math.max(0L, _)` on the lower part for `innerProduct` as well.

What failures are you seeing when you run those tests?  Are they expecting approximations with negative lower bounds?

**Comment by miguno** [comment]
The error I am seeing looks as follows:

```
[info] CMSBigIntTest:
[info] A Count-Min sketch implementing CMSCounting
...snip...
[info] - should estimate inner products *** FAILED ***
[info]   32743 was greater than or equal to 0, but 32743 was not less than or equal to 27394 (CountMinSketchTest.scala:215)
...snip...
```

In this specific example ouput of the failure, the generated `Approximate` instance was:

```scala
Approximate(6396, 33790, 33790, 0.9999999999)
```

with associated local test variables (see test code below):

```
approx          = 33790 = Approximate.estimate
exact           =  1047
estimationError = 32743 = (approx - exact)
maxError        = 27394 = (approx - Approximate.min)
```

I started digging into negative minimums of `Approximate` because I noticed that all the other CMS variants for which the tests passed (`CMS[Long]`, `CMS[Int]`, `CMS[Short]`) showed negative values whereas the `CMS[BigInt]` test case (when using the updating BigInt-hashing code mentioned in #392) showed a positive value.  So it was but one path that I found worthy to pursue while debugging.

> Also: In general I noticed a big discrepancy between the approximated and the exact inner product during these tests.  The exact inner product is typically ~1,200, whereas the successful estimates (current hashing = Long,Int,Short) are ~25,000 and the failed estimates (Murmur3 hashing = BigInt) are ~32,000.  This might be an artifact of the CMS parameters and/or input data patterns used for the inner product test case.  I can't say right now.

The last line in the test spec for inner products triggers the failure:

```scala
    "estimate inner products" in {
      val totalCount = 5234
      val range = 1390
      val data1 = (0 to (totalCount - 1)).map { _ => RAND.nextInt(range) }.toK[K]
      val data2 = (0 to (totalCount - 1)).map { _ => RAND.nextInt(range) }.toK[K]
      val cms1 = COUNTING_CMS_MONOID.create(data1)
      val cms2 = COUNTING_CMS_MONOID.create(data1)

      val approxA = cms1.innerProduct(cms2)
      val approx = approxA.estimate
      val exact = exactInnerProduct(data1, data2)
      val estimationError = approx - exact
      val maxError = approx - approxA.min
      val beWithinTolerance = be >= 0L and be <= maxError

      approx should be(cms2.innerProduct(cms1).estimate)
      approx should be >= exact
      estimationError should beWithinTolerance   // <<< this line triggers the failure
    }
```

**Comment by miguno** [comment]
And another surprise:  It looks as if the inner product test passes for the current CMS code only by sheer luck. :-)

When I _lower_ the `totalCount` value to, say, `2000`, then the inner product test begins to fail indiscriminately for all CMS variants (i.e. Long, Int, Short, BigInt).

```scala
"estimate inner products" in {
      val totalCount = 2000 // instead of currently: 5234
```

results in (here: `CMS[Int]` test spec):

```
4767 was greater than or equal to 0, but 4767 was not less than or equal to 4000
```

When using `totalCount = 3190`, for example, the test may pass or may fail.  Yes, we do generate random input but nonetheless the mathematical properties of CMS should not be violated.

Similarly, if I _increase_ `totalCount` to, say, `>= 20,000` the test will also pass occasionally for the modified `CMS[BigInt]` code.

In other words: The likelihood of test success seems to increase and decrease with the size of `totalCount`.

**Comment by miguno** [comment]
PS: If this turns out to be a legitimate issue, then we should probably use a ScalaCheck-based property test for estimating inner products instead of a single test fixture.

**Comment by jnievelt** [comment]
It seems the failure you're seeing stems from the higher estimate, as the `eps * totalCount * totalCount` should come out the same for any key type.  The primary suspect would be a hashing algorithm which collides more often than the others.

Have you looked at how these number come out against the existing BigInt hashing?  It might also be useful to similarly compare approximations coming back from `frequency`, though I imagine the discrepancies might not be quite as pronounced.

**Comment by miguno** [comment]
I finally found and solve the issue.  There were bugs in the code as well as in the tests (I spend a lot of time figuring why the fixed code supposedly wasn't working until I realized the tests were broken).  I will sent a pull request via #392 shortly.
